msgp/write.go

```diff
@@ -120,16 +125,25 @@ func NewWriter(w io.Writer) *Writer {
 
 // NewWriterSize returns a writer with a custom buffer size.
 func NewWriterSize(w io.Writer, sz int) *Writer {
-	// we must be able to require() 18
+	// we must be able to require() 'minWriterSize'
 	// contiguous bytes, so that is the
 	// practical minimum buffer size
-	if sz < 18 {
-		sz = 18
+	if sz < minWriterSize {
+		sz = minWriterSize
 	}
+	buf := make([]byte, sz)
+	return NewWriterBuf(w, buf)
+}
 
+// NewWriterBuf returns a writer with a provided buffer.
+// It panics if the buffer size is smaller than 'minWriterSize'.
+func NewWriterBuf(w io.Writer, buf []byte) *Writer {
+	if len(buf) < minWriterSize {
```

Comment:
In principle you can stretch `buf` to `buf[:cap(buf)]` here.

But I think it would be fine to reallocate the buffer if its capacity is actually under 18 bytes, provided that the API and docs convey that information appropriately.

Reply:
Updated